packages/react-sdk/src/hooks/useCachedMessages.ts

```diff
@@ -0,0 +1,260 @@
+import type { Client, Conversation, ListMessagesOptions } from "@xmtp/xmtp-js";
+import { DecodedMessage, SortDirection } from "@xmtp/xmtp-js";
+import { useCallback, useEffect, useRef, useState } from "react";
+import type { Collection, IndexableType } from "dexie";
+import type { CachedMessage } from "../helpers/messagesDb";
+import { messagesDb } from "../helpers/messagesDb";
+import { getConversationId } from "../helpers/getConversationId";
+import { useClient } from "./useClient";
+import { updateLastEntry } from "../helpers/updateLastEntry";
+
+const filterByConversation = (cId: string) => (cachedMessage: CachedMessage) =>
+  cachedMessage.cId === cId;
+
+const fastForward = (lastEntry: DecodedMessage | undefined, cId: string) => {
+  let fastForwardComplete = false;
+  return (item: CachedMessage) => {
+    if (fastForwardComplete) {
+      return item.cId === cId;
+    }
+    if (lastEntry !== undefined && item.id === lastEntry.id) {
+      fastForwardComplete = true;
+    }
+    return false;
+  };
+};
+
+type GetCachedMessagesOptions = {
+  /**
+   * Conversation ID
+   */
+  cId: string;
+  /**
+   * XMTP client, needed for decryption
+   */
+  client: Client;
+  /**
+   * Sort direction
+   */
+  direction?: string;
+  /**
+   * Is this the initial fetch?
+   */
+  initial?: boolean;
+  /**
+   * The position of the cursor
+   */
+  lastEntry?: DecodedMessage;
+  /**
+   * Page size
+   */
+  limit?: number;
+};
+
+// fetch messages from the cache and decode them
+const getCachedMessages = async ({
+  cId,
+  client,
+  direction,
+  initial,
+  lastEntry,
+  limit,
+}: GetCachedMessagesOptions) => {
+  let messagesQuery: Collection<CachedMessage, IndexableType>;
+
+  if (initial) {
+    // the initial query is slightly different than subsequent queries
+    messagesQuery = messagesDb.messages
+      .orderBy("sent")
+      .filter(filterByConversation(cId));
```

Comment:
Why isn't this just an additional where clause? This will fetch _all_ records in the DB and then filter them in your app, instead of using the DB index?

Reply:
i wish this was possible, but it seems like you can't combine `where` and `orderBy`. i mostly followed the examples in the docs.

see [here](https://dexie.org/docs/Table/Table.orderBy()) and [here](https://dexie.org/docs/Collection/Collection.offset()#paged-or-queries)

Reply:
I wonder if you can re-jig the schema to have a compound index for `[cid+sent]`. I would _think_ that would get around the limitation.

Reply:
the current Dexie API is not that great for working with compound indexes. finally figured out that `between` can work well. updated the query.

[Refactor DB querying](https://github.com/xmtp/xmtp-web/pull/37/commits/e34e75da973bd105d3ab455e9b3d7ca011bea775)